server/src/main/java/org/opensearch/search/aggregations/metrics/SumAggregatorFactory.java

```diff
@@ -69,11 +66,24 @@ static void registerAggregators(ValuesSourceRegistry.Builder builder) {
         builder.register(
             SumAggregationBuilder.REGISTRY_KEY,
             List.of(CoreValuesSourceType.NUMERIC, CoreValuesSourceType.DATE, CoreValuesSourceType.BOOLEAN),
-            SumAggregator::new,
+            SumAggregatorFactory::parseSumAggregator,
             true
         );
     }
 
+    private static Aggregator parseSumAggregator(String s, ValuesSourceConfig valuesSourceConfig, SearchContext searchContext, Aggregator aggregator, Map<String, Object> stringObjectMap) throws IOException {
+        if (!valuesSourceConfig.hasValues()) {
+            return new SumNullAggregator(s, valuesSourceConfig, searchContext, aggregator, stringObjectMap);
+        }
+
+        // use kahan summation for floating point numbers
+        if (((ValuesSource.Numeric) valuesSourceConfig.getValuesSource()).isFloatingPoint()) {
```

Comment:
It seems like there is interesting by product of this branching logic: the aggregation across different indices that have same field but different types (let say `long` and `double`) does not produce the correct results anymore:

index1:
```
        "mappings": {
            "properties": {
                "counter": { "type": "double" }
            }
        }
```

index2:
```
        "mappings": {
            "properties": {
                "counter": { "type": "long" }
            }
        }
```

/_bulk:
```
{"index": {"_index": "index1", "_id": "1"}}
{ "value" :-9223372026854775808.0 }
{"index": {"_index": "index1", "_id": "2"}}
{ "value" : 1000.0 }
{"index": {"_index": "index2", "_id": "3"}}
{ "value" : 9223372026854775807 }
{"index": {"_index": "index2", "_id": "4"}}
{ "value" : -500 }
```

index*/_search:
```
{
  "query": {
      "match_all": {}
  },
  "aggs": {
    "counters": {
        "sum": {
           "field": "counter"
        }  
    }
  }
}
```

The results are surprising:

```
    "aggregations": {
        "counters": {
            "value": 0.0
        }
    }
```

Reply:
This made me think of another issue with the current implementation in that it doesn't fix the incorrect summation of double values, because only long values are treated as different.

I do have one idea for something that would solve this issue as well as allow this solution to work with all data types but it is a little abstract and might not be efficient:
1. Split every value to be summed into a `biginteger` and a `double` containing the floating point numbers
2. Sum every `biginteger` together which will keep the precision when adding and subtracting large values
3. Sum all floating point `double` values together
4. Convert the `biginteger` into a double and add the floating point back onto the end

Reply:
Do you know if we could just use `BigDecimal` instead? Also, most of the counting aggregators (AvgAggregator, ...) should be changed apparently.

Reply:
Sadly it appears that BigDecimal has the same loss of precision problem that normal double arithmetic faces

Reply:
@reta I just had a look through this test code again and I believe that there is actually a mistake in it, it appears that you are calling the sum of "counter" rather than "value". I find that if I update it to say "value" I actually receive the correct response outputted. When you get a chance do you think that you could run the test again and see if that solves the issue?

Reply:
I came back to this and noticed something interesting, I believe there may be a mistake in the schema that you posted. When I try to run the commands that you run the result is always 0 regardless of which version of opensearch you use, I believe the issue is that in:
```
{
  "query": {
      "match_all": {}
  },
  "aggs": {
    "counters": {
        "sum": {
           "field": "counter"
        }  
    }
  }
}
```

`counter` should be `value`, if I make that change I then start receiving results that make sense. Although the results are not completely correct because of the mixing of double and long values which causes a loss of accuracy, I do believe that the loss of accuracy is the same as base OpenSearch. When working with just long values it appears that the accuracy problem has been completely solved, I will do more research into this solution to confirm that it fully works but it currently looks very promising.